subprojects/maven/src/main/java/org/gradle/api/publish/maven/tasks/PublishToMavenRepository.java

```diff
@@ -18,20 +18,29 @@
 
 import org.gradle.api.InvalidUserDataException;
 import org.gradle.api.artifacts.repositories.MavenArtifactRepository;
+import org.gradle.api.credentials.Credentials;
+import org.gradle.api.provider.Property;
 import org.gradle.api.publish.internal.PublishOperation;
 import org.gradle.api.publish.maven.internal.publication.MavenPublicationInternal;
 import org.gradle.api.publish.maven.internal.publisher.MavenPublisher;
 import org.gradle.api.publish.maven.internal.publisher.ValidatingMavenPublisher;
+import org.gradle.api.tasks.Input;
 import org.gradle.api.tasks.Internal;
 import org.gradle.api.tasks.TaskAction;
+import org.gradle.internal.artifacts.repositories.AuthenticationSupportedInternal;
 
 /**
  * Publishes a {@link org.gradle.api.publish.maven.MavenPublication} to a {@link MavenArtifactRepository}.
  *
  * @since 1.4
  */
 public class PublishToMavenRepository extends AbstractPublishToMaven {
+
+    private static final Credentials NO_CREDENTIALS = new Credentials() {
```

Comment:
Same here, I would use an actual missing value to mean "no credentials" and mark `getCredentials()` with `@Optional`.